File: csharp/BronKerbosch/GraphDegeneracy.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;

namespace BronKerbosch
{
    public class Degeneracy
    {
        // Iterate connected vertices, lowest degree first.
        // drop=N: omit last N vertices
        public static IEnumerable<Vertex> Ordering(UndirectedGraph graph, int drop)
        {
            Debug.Assert(drop >= 0);
            const int NO_PRIORITY = -1;
            int[] priorityPerVertex = new int[graph.Order];
            var max_priority = 0;
            var numLeftToPick = 0;
            foreach (var i in Enumerable.Range(0, graph.Order))
            {
                var c = new Vertex(i);
                var degree = graph.Degree(c);
                if (degree > 0)
                {
                    var priority = degree;
                    max_priority = Math.Max(max_priority, priority);
                    priorityPerVertex[i] = degree;
                    numLeftToPick += 1;
                }
            }
            // Possible values of priority_per_vertex:
            //   no_priority: when yielded or if unconnected
            //   0..max_priority: candidates still queued with priority (degree - #of yielded neighbours)
            var q = new PriorityQueue(max_priority);
            foreach (var (c, p) in priorityPerVertex.Select((p, i) => (new Vertex(i), p)))
            {
                if (p > 0)
                    q.Put(priority: p, element: c);
            }

            numLeftToPick -= drop;
            while (numLeftToPick >= 0)
            {
                numLeftToPick -= 1;
                var i = q.Pop();
                while (priorityPerVertex[i] == NO_PRIORITY)
                {
                    // was requeued with a more urgent priority and therefore already picked
                    i = q.Pop();
                }
                Debug.Assert(priorityPerVertex[i] >= 0);
                priorityPerVertex[i] = NO_PRIORITY;
                yield return i;
                foreach (Vertex v in graph.Neighbours(i))
                {
                    var oldPriority = priorityPerVertex[v];
                    if (oldPriority != NO_PRIORITY)
                    {
                        // Since this is an unvisited neighbour of a vertex just being picked,
                        // its priority can't be down to the minimum.
                        Debug.Assert(oldPriority > 0);
                        // Requeue with a more urgent priority, but don't bother to remove
                        // the original entry - it will be skipped if it's reached at all.
                        priorityPerVertex[v] = oldPriority - 1;
                        q.Put(priority: oldPriority - 1, element: v);
                    }
                }
            }
        }
    }

    class PriorityQueue
    {
        private List<Vertex>[] queuePerPriority;

        public PriorityQueue(int max_priority)
        {
            queuePerPriority = new List<Vertex>[max_priority + 1];
        }

        public void Put(int priority, Vertex element)
        {
            Debug.Assert(priority >= 0);
            if (queuePerPriority[priority] == null)
                queuePerPriority[priority] = new List<Vertex>();
            queuePerPriority[priority].Add(element);
        }

        public Vertex Pop()
        {
            foreach (var queue in queuePerPriority)
            {
                if (queue != null)
                {
                    var last = queue.Count - 1;
                    if (last >= 0)
                    {
                        var v = queue[last];
                        queue.RemoveAt(last);
                        return v;
                    }
                }
            }
            throw new ArgumentException("Cannot pop more than has been put");
        }
    }
}

File: go/src/BronKerbosch/graph_degeneracy.go
package BronKerbosch

type VertexVisitor interface {
	visit(Vertex)
	Close()
}

type SimpleVertexVisitor struct {
	vertices []Vertex
}

type ChannelVertexVisitor struct {
	vertices chan<- Vertex
}

func (g *SimpleVertexVisitor) visit(v Vertex) {
	g.vertices = append(g.vertices, v)
}

func (g *SimpleVertexVisitor) Close() {
}

func (g *ChannelVertexVisitor) visit(v Vertex) {
	g.vertices <- v
}

func (g *ChannelVertexVisitor) Close() {
	close(g.vertices)
}

func degeneracyOrdering(graph *UndirectedGraph, visitor VertexVisitor, drop int) {
	if drop > 0 {
		panic("expecting negative drop value")
	}
	defer func() { visitor.Close() }()
	order := graph.order()
	// Possible values of priorityPerNode:
	//   -1: when yielded
	//   0..maxDegree: candidates still queued with priority (degree - #of yielded neighbours)
	priorityPerNode := make([]int, order)
	maxDegree := 0
	numLeftToVisit := 0
	for c := 0; c < order; c++ {
		degree := graph.degree(Vertex(c))
		if degree > 0 {
			priorityPerNode[Vertex(c)] = degree
			if maxDegree < degree {
				maxDegree = degree
			}
			numLeftToVisit++
		}
	}
	numLeftToVisit += drop
	if numLeftToVisit <= 0 {
		return
	}

	var q priorityQueue
	q.init(maxDegree)
	for c, p := range priorityPerNode {
		if p > 0 {
			q.put(p, Vertex(c))
		}
	}

	for {
		i := q.pop()
		for priorityPerNode[i] == -1 {
			// was requeued with a more urgent priority and therefore already visited
			i = q.pop()
		}

		visitor.visit(i)
		numLeftToVisit--
		if numLeftToVisit == 0 {
			return
		}

		priorityPerNode[i] = -1
		for v := range graph.adjacencies[i] {
			p := priorityPerNode[v]
			if p != -1 {
				// Requeue with a more urgent priority, but don't bother to remove
				// the original entry - it will be skipped if it's reached at all.
				priorityPerNode[v] = p - 1
				q.put(p-1, v)
			}
		}
	}
}

type priorityQueue struct {
	stackPerPriority [][]Vertex
}

func (q *priorityQueue) init(maxPriority int) {
	q.stackPerPriority = make([][]Vertex, maxPriority+1)
}

func (q *priorityQueue) put(priority int, element Vertex) {
	q.stackPerPriority[priority] = append(q.stackPerPriority[priority], element)
}

func (q *priorityQueue) pop() Vertex {
	for p := 0; ; p++ {
		l := len(q.stackPerPriority[p])
		if l > 0 {
			last := q.stackPerPriority[p][l-1]
			q.stackPerPriority[p] = q.stackPerPriority[p][:l-1]
			return last
		}
	}
}
